Include block signatures in periodic catchup

hewison-chris opened the pull request:
Fixes issue #1635

kchulj commented:
Could this fix this bug during the catch-up?
```
2021-02-17 19:57:09,472 Error [agora.node.Validator] - Block failed to validate: Block: Height is not one more than previous block
```

hewison-chris commented:
> Could this fix this bug during the catch-up?
> 
> ```
> 2021-02-17 19:57:09,472 Error [agora.node.Validator] - Block failed to validate: Block: Height is not one more than previous block
> ```

Unfortunately I don't think it will.

Geod24 commented on source/agora/api/FullNode.d:
```diff
@@ -292,4 +292,19 @@ public interface API
+            Block header signature with info of validators who signed
+
+    ***************************************************************************/
+
```
I think we should have an API for the `BlockHeader` instead of a custom entry like this, it'd be more generic.

Geod24 commented on source/agora/network/NetworkClient.d:
```diff
@@ -369,6 +369,22 @@ class NetworkClient
+
+    ***************************************************************************/
+
+    public BlockHeaderSignature[] getBlockSignature (ulong height) @trusted nothrow
```
Why trusted ?

hewison-chris replied:
Same as other functions in the NetworkClient as `attemptRequest` is not `@safe`.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -572,14 +572,16 @@ extern(D):
             if (!this.collectBlockSignature(block_sig, block.hashFull()))
                 return;
-            const signed_block = updateMultiSignature(block);
+            auto signed_block = updateMultiSignature(block);
```
You can leave it `const` if you assign with a `foreach` (and pass a `const` instance of `BlockHeaderSignature` around).

Geod24 commented on source/agora/network/NetworkManager.d:
```diff
@@ -916,6 +916,52 @@ public class NetworkManager
+                        }
+                        catch (Exception e)
+                        {
+                            log.error("getMissingBlockSigs: Exception during updating block signature at height {}: {}", height, e.msg);
```
Is that proper English ? "during updating"

hewison-chris replied:
Too long away from homeland.
Corrected.

Geod24 commented on source/agora/node/BlockStorage.d:
```diff
@@ -977,19 +966,18 @@ public class MemBlockStorage : IBlockStorage
+    public override bool updateBlockMultiSig (BlockHeaderSignature blockHeaderSignature) @safe
     {
-        if (this.blocks.length < block.header.height)
-            return false;
```
Why removing this check ?

hewison-chris replied:
Updated

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -867,4 +868,26 @@ public class FullNode : API
         return found_txs;
     }
+
+    /***************************************************************************
+
+        Params:
+            height = Height of block
+
+        Returns:
+            BlockHeaderSignature if the client has that block
+
+    ***************************************************************************/
+
+    public BlockHeaderSignature[] getBlockSignature (ulong height) @safe
+    {
+        BlockHeaderSignature[] block_signatures;
+        if (height <= this.getBlockHeight())
+        {
+            auto block = this.ledger.getBlockAtHeight(Height(height));
+            block_signatures = [ BlockHeaderSignature(Height(height), block.header.signature,
+                block.header.validators) ];
+        }
+        return block_signatures;
+    }
```
Why does it return an array if we only ever return one value ?

hewison-chris replied:
Now it return multiple values.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -842,6 +842,34 @@ public class Ledger
             .map!(idx => this.storage.readBlock(Height(idx)));
     }
 
+    /***************************************************************************
+
+        Get a block at the provided block height.
+
+        Params:
+            height = the block height
+
+        Returns:
+            the block
+
+    ***************************************************************************/
+
+    public Block getBlockAtHeight (Height height) @safe nothrow
+    {
+        if (height == this.last_block.header.height)
+            return this.last_block;
+        else
+            try
+            {
+                return this.storage.readBlock(Height(height));
+            }
+            catch(Exception e)
+            {
+                log.fatal("Block failed to be loaded from storage for height {}: {}", height, e);
+                assert(0);
+            }
+    }
+
```
Nope, let's not add more storage dependence to the ledger. Additionally, no `assert(0)`

hewison-chris replied:
Removed as not needed anymore.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1036,6 +1064,35 @@ public class Ledger
     {
         return this.slash_man.getRandomSeed(this.last_block.header.height);
     }
+
+    /***************************************************************************
+
+        Count the active validators
+
+        Returns:
+            the count of enrolled validators for given height
+
+    ***************************************************************************/
+
+    public size_t getCountOfValidatorsAtBlockHeight(Height height) nothrow @safe
+    {
+        return this.enroll_man.getCountOfValidators(height);
+    }
+
+    /***************************************************************************
+
+        Find the count of validators who signed the block for given height
+
+        Returns:
+            the range of missing validator indices
+
+    ***************************************************************************/
+
+    public size_t getCountOfValidatorsSignedBlockSig(Height height) nothrow @safe
+    {
+        return iota(0, this.enroll_man.getCountOfValidators(height)).filter!(i =>
+            this.getBlockAtHeight(height).header.validators[i]).count();
+    }
```
Those are very specific, single use functions.

hewison-chris replied:
Removed.

Geod24 commented on tests/unit/BlockStorageMultiSig.d:
```diff
@@ -64,40 +64,38 @@ private void main ()
 
-    iota(1, BlockCount).each!(h => {
-        // Sign each block with signature and set validator 1 as signed
+    void signBlockSig1 (Height h)
+    {
```
👍

Geod24 reviewed (changes requested):
This won't work for any kind of overlap, will it ?

hewison-chris committed:
Include block signatures during periodic catchup.

For last last 3 blocks if not all signatures are included in the block
header then it will attempt to find block headers with more signatures
from connected peers and then update if it finds them.

hewison-chris commented:
Converted back to draft after rebase from master and started addressing the comments. Not yet ready to be reviewed again.

hewison-chris commented:
All 🟢

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -1062,6 +1062,7 @@ public class Ledger
     {
         return this.slash_man.getRandomSeed(this.last_block.header.height);
     }
+
```
? :O

Geod24 commented on tests/unit/BlockStorageMultiSig.d:
```diff
@@ -65,40 +65,37 @@ private void main ()
 
-    iota(1, BlockCount).each!(h => {
-        // Sign each block with signature and set validator 1 as signed
+    void signBlockSig1 (Height h)
+    {
```
Own commit ?

Geod24 commented on source/agora/test/PeriodicCatchup.d:
```diff
@@ -41,12 +43,24 @@ private extern(C++) class DoesNotExternalizeBlockNominator : TestNominator
-        if (slot_idx == 2)
+        try
         {
-            log.trace("Do not externalize this block to test periodic catchup");
-            return;
+            if (slot_idx == 2)
+            {
+                log.trace("Remove signatures for block 1 to test signature catchup");
+                BlockHeader header = super.ledger.getBlocksFrom(Height(1)).front.header;
+                header.validators = BitField!ubyte(6);
+                header.signature = Signature.init;
+                super.ledger.updateBlockMultiSig(header);
+                log.trace("Do not externalize block 2 to test periodic catchup of blocks");
+                return;
+            }
+            super.valueExternalized(slot_idx, value);
+        }
+        catch (Exception e)
+        {
+            log.error("PeriodicCatchup exception thrown during test: {}" , e);
         }
```
That looks like an error, why not let it propagate, or at least `assert(0)` ?

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -328,19 +328,19 @@ public class Ledger
 
         Params:
-            height = height of block to be updated
+            BlockHeader = block header to be updated
```
```suggestion
            header = block header to be updated
```

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -867,4 +868,31 @@ public class FullNode : API
+
+    ***************************************************************************/
+
+    public BlockHeader[] getBlockHeaders (Set!ulong heights) @safe
+    {
+        import std.algorithm: min;
+        import std.conv;
+
+        BlockHeader[] headers;
+        if (!heights.empty)
+        {
+            foreach (block; this.ledger.getBlocksFrom(Height(heights[].front)))
+            {
+                if (block.header.height in heights)
+                    headers ~= block.header;
+            }
+        }
+        return headers;
+    }
```
This is an expensive API. We really need to improve it, but let's go with it for now.

Geod24 reviewed (approved):
Just a few changes

hewison-chris committed:
Unit test:use inner function to improve compilation failure message
